doc/pyprinttags.1
.TH pyprinttags 1 "20 March 2016" "version 1.3.0"
.\" Copyright 2013-2016 Michael Helmling.
.\" Copying permitted according to the GNU General Public License V3.


.SH NAME
pyprinttags - print tags of audio files to stdout

.SH SYNOPSIS
.SY pyprinttags
.OP -b|--batch
.I
file1 [file2] ... 
.SY

.SH DESCRIPTION
.B pyprinttags
is a small script that comes with the 
.BR pytaglib " python tag reading library. It prints all tags found in the audio files given as command line
parameters.
.PP
If a file contains metadata that can not be textually represented, this is indicated in the output as well, and the
program gives asks whether to remove those pieces from the file. That behavior can be disabled by the
.B -b
option.

.SH OPTIONS
.TP
.B -b, --batch
disable user interaction in case of unsupported metadata

.SH AUTHOR
.MT michaelhelmling@posteo.de
Michael Helmling 
.ME
.SH AVAILABILITY
The newest version of
.B pytaglib
containing 
.B pyprinttags
can be always obtained from the package's home page
.UR https://pypi.python.org/pypi/pytaglib
.UE


NEWS
Changelog
=========

pytaglib 1.3.0 (2016/07/22)
---------------------------
- Remove workaround for pre-1.9 taglib versions in order to reduce codebase
- code cleanup
- move tests out of source folder

pytaglib 1.2.1 (2016/07/17)
---------------------------
This is a non-feature release (no change to the code base)
- update README
- build windows wheel agains taglib 1.11

pytaglib 1.2.0 (2016/03/20)
---------------------------
- add Windows support (see README)
- update copyright dates
- update cython version used to build shipped taglib.cpp to 0.23.4

pytaglib 1.1.0 (2015/09/06)
---------------------------
- add a `File.close()` method that ends all I/O operations.

pytaglib 1.0.3 (2015/03/16)
---------------------------
- include ReST version of the README for pypi (converted using pandoc)

pytaglib 1.0.2 (2015/03/15)
---------------------------
- ensure that pyprinttags removes unsupported properties only when user enters 'y' or 'yes'
  (thanks to lahwaacz)
- fix a typo in the readme (thanks to panzl)

pytaglib 1.0.1 (2015/03/09)
---------------------------
- cleaned up source code and made it more readable
- update README to contain more information
- no functional API changes

pytaglib 1.0 (2015/01/03)
-------------------------
- as the library has been used for several years now without any known
  bugs, I declare it as stable.
- cleaned up source. Especially simplified several statements due to improvements
  in recent Cython versions.
- The workaround for MPEG files with taglib <= 1.8 is now forced disabled if taglib version >= 1.9
  is detected.
- update copyright dates

pytaglib 0.4 (2014/03/29)
-------------------------
- remove Cython dependency by shipping taglib.cpp

pytaglib 0.3.7 (2014/01/21)
---------------------------
- remove a test file that looked a little non-free
- update copyright dates

pytaglib 0.3.6 (2013/08/13)
---------------------------
- fix Python 2.6 support in pyprinttags
- update copyright dates

pytaglib 0.3.5 (2013/04/03)
---------------------------
- add support for Python 2.6 by replacing some methods added in 2.7

pytaglib 0.3.4 (2013/01/16)
---------------------------
- move cython from install_requires to setup_requires

pytaglib 0.3.3 (2013/01/16)
---------------------------
- ensure sources are included in sdist packages 
- fix call to pyprinttags
- fix setup.py handling for non-utf8-locales
- rename pyprinttags to pyprinttags3 for python3 installs

pytaglib 0.3.2 (2013/01/12)
---------------------------
- add "batch mode" to pyprinttags and allow several files at once
- add a man page for pyprinttags
- remove .travis.yml since travis' build system is too old for building pytaglib

pytaglib 0.3.1 (2013/01/07)
---------------------------
- updated packaging information in setup.py

pytaglib 0.3.0 (2012/12/14)
---------------------------
- implement a hack that works around a bug in taglib, leading
  to several problems in connection with MP3 files with ID3v1
  tags. Open files with ``f=taglib.File(path, applyID3v2Hack=True)``
  to ensure that MPEG files will always get updated to ID3v2 if
  necessary. Use at your own risk!
- update documentation
- declare development stadium as "beta" since no critical bugs
  seem to exist.

pytaglib 0.2.5 (2012/12/05)
---------------------------
- fix integration into PyPI, clean up code & documentation

pytaglib 0.2.4 (2012/09/09)
---------------------------
- add test script for python2.x/3.x

pytaglib 0.2.3 (2012/07/30)
---------------------------

- save() now returns unsuccessful tags due to metadata format
- add taglib.version attribute to get module version
- all tests pass with my taglib fork
- fix Python2 compatibility
- README now in ReSt format
- code cleanup and documentation improvement

pytaglib 0.2.2 (2012/07/25)
---------------------------

- Switch to setuptools/distribute, prepare publishing on PyPI
- Add some basic unit tests (test files stolen from taglib)
- Ensure package works with Python2.x/3.x


src/pyprinttags.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Copyright 2011-2016 Michael Helmling, michaelhelmling@posteo.de
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 3 as
# published by the Free Software Foundation
#
"""A sample script printing the tags of a given audio file.

The main purpose is to show how pytaglib is used, but it also serves as a tool
showing *all* metadata of a given while, while most taggers only display a set
of certain tags they know.
"""
from __future__ import unicode_literals, print_function

import argparse
import os.path
import sys
import taglib


def script():
    """Run the command-line script."""
    parser = argparse.ArgumentParser(description="Print all textual tags of one or more audio files.")
    parser.add_argument("-b", "--batch", help="disable user interaction", action="store_true")
    parser.add_argument("file", nargs="+", help="file(s) to print tags of")
    args = parser.parse_args()
    for filename in args.file:
        if isinstance(filename, bytes):
            filename = filename.decode(sys.getfilesystemencoding())
        line = "TAGS OF '{0}'".format(os.path.basename(filename))
        print("*" * len(line))
        print(line)
        print("*" * len(line))
        audioFile = taglib.File(filename)
        tags = audioFile.tags
        if len(tags) > 0:
            maxKeyLen = max(len(key) for key in tags.keys())
            for key, values in tags.items():
                for value in values:
                    print(('{0:' + str(maxKeyLen) + '} = {1}').format(key, value))
        if len(audioFile.unsupported) > 0:
            print('Unsupported tag elements: ' + "; ".join(audioFile.unsupported))
            if sys.version_info[0] == 2:
                inputFunction = raw_input
            else:
                inputFunction = input
            if not args.batch and inputFunction("remove unsupported properties? [yN] ").lower() in ["y", "yes"]:
                audioFile.removeUnsupportedProperties(audioFile.unsupported)
                audioFile.save()

if __name__ == '__main__':
    script()
